pkg/policy/manager.go

```diff
@@ -0,0 +1,455 @@
+		if err != nil {
+			return fmt.Errorf("failed to open repository: %s", err.Error())
+		}
+	default:
+		return fmt.Errorf("failed to clone repository: %s", err.Error())
+	}
+
+	return nil
+}
+
+// RunPolicy runs the given policy.
+func (m *ManagerImpl) RunPolicy(ctx context.Context, execReq *ExecuteRequest) (*ExecutionResult, error) {
+	p := execReq.Policy
+
+	// Check if given policy exists in our policy folder
+	osFs := file.NewOsFs()
+	orgPolicyStr := filepath.Join(p.Organization, p.Repository)
+	repoFolder := filepath.Join(m.policyDirectory, defaultLocalSubPath, orgPolicyStr)
+	if info, err := osFs.Stat(repoFolder); err != nil || !info.IsDir() {
+		return nil, fmt.Errorf("could not find policy '%s' locally. Try to download the policy first", orgPolicyStr)
+	}
+
+	// If repository path was specified, also check if that exists
+	policyFolder := repoFolder
+	if p.RepositoryPath != "" {
+		policyFolder = filepath.Join(repoFolder, p.RepositoryPath)
+		if info, err := osFs.Stat(policyFolder); err != nil || !info.IsDir() {
+			return nil, fmt.Errorf("could not find policy '%s' in the folder '%s'. Try to download the policy first", orgPolicyStr, p.RepositoryPath)
+		}
+	}
+
+	// Checkout policy repository tag
+	if err := p.checkoutPolicyVersion(repoFolder); err != nil {
+		return nil, fmt.Errorf("failed to checkout repository tag: %s", err.Error())
+	}
+
+	// Make sure policy file exists
+	var policyFilePath string
+	for _, extensionName := range defaultSupportedPolicyExtensions {
+		currPolicyFile := filepath.Join(policyFolder, fmt.Sprintf("%s.%s", defaultPolicyFileName, extensionName))
+		if _, err := osFs.Stat(currPolicyFile); err == nil {
+			policyFilePath = currPolicyFile
+			break
+		}
+	}
+	if policyFilePath == "" {
+		return nil, fmt.Errorf("failed to find policy file; policy.%#v not found in %s", defaultSupportedPolicyExtensions, policyFolder)
+	}
+
+	// Read policy file
+	parser := config.NewParser(nil)
+	policiesRaw, diags := parser.LoadHCLFile(policyFilePath)
+	if diags != nil && diags.HasErrors() {
+		return nil, fmt.Errorf("failed to load policy file: %#v", diags.Errs())
+	}
+	policies, diagsDecode := parser.DecodePolicies(policiesRaw, diags)
+	if diagsDecode != nil && diagsDecode.HasErrors() {
+		return nil, fmt.Errorf("failed to parse policy file: %#v", diagsDecode.Errs())
+	}
+
+	// Acquire connection from the connection pool
+	conn, err := m.pool.Acquire(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to acquire connection from the connection pool: %s", err.Error())
+	}
+	defer conn.Release()
+
+	// Traverse all policies recursively
+	policyMap := make(map[string]*config.Policy)
+	policyMap = traversePolicies(policies.Policies, "", policyMap)
+
+	// No sub path provided. Execute everything.
+	executor := NewExecutor(conn)
+	if p.SubPath == "" {
+		return executor.ExecutePolicies(ctx, execReq, policyMap)
+	}
+
+	// If we are here, the user only wants to execute a sub policy/view/query so we have to
```

Comment:
suggestion: wouldn't it be better to split this function to the 3 functions -> normal execution, sub policy execution, single query execution?

Reply:
I agree. Sounds reasonable.